File: LabProject/src/main/java/agh/ics/oop/Animal.java
package agh.ics.oop;

import java.util.ArrayList;
import java.util.HashMap;

public class Animal implements IMapElement{
    private MapDirection direction;
    private IWorldMap map;
    private Vector2d position;

    private ArrayList<IPositionChangeObserver> observers = new ArrayList();
    public Animal(IWorldMap map) {this(map, new Vector2d(2,2));}

    public Animal(IWorldMap map, Vector2d initialPosition)
    {
        this.map = map;
        this.position = initialPosition;
        this.direction = MapDirection.NORTH;
    }

    @Override
    public String toString() {
        return switch (this.direction){
            case EAST -> "E";
            case WEST -> "W";
            case NORTH -> "N";
            case SOUTH -> "S";
        };

    }

    public Vector2d getPosition() {
        return position;
    }

    public MapDirection getDirection() {
        return direction;
    }

    public boolean isAt(Vector2d position){
        return (this.position.equals(position));
    }

    public void addObserver(IPositionChangeObserver observer){
        observers.add(observer);
    }

    public void removeObserver(IPositionChangeObserver observer){
        observers.remove(observer);
    }

    private void positionChanged(Vector2d oldposition, Vector2d newposition){
        for(IPositionChangeObserver observer:observers){
            observer.positionChanged(oldposition,newposition);
        }
    }
    public void move(MoveDirection direction){
        switch(direction) {
            case RIGHT -> this.direction = this.direction.next();
            case LEFT -> this.direction = this.direction.previous();
            default -> {
                Vector2d directionVector = this.direction.toUnitVector();
                if (direction == MoveDirection.BACKWARD) {
                    directionVector = directionVector.opposite();
                }
                Vector2d newLocation = this.position.add(directionVector);
                if (this.map.canMoveTo(newLocation)) {
                    positionChanged(this.position, newLocation);
                    this.position = newLocation;
                }}}
    }

}

File: KotlinLab/src/main/kotlin/Animal.kt
class Animal (
    var direction: MapDirection = MapDirection.NORTH,
    private val map: IWorldMap? = null,
    var position: Vector2d = Vector2d(2.0,2.0)
            //private set - how??
        )
{
    override fun toString(): String = when (direction) {
            MapDirection.EAST -> "E"
            MapDirection.WEST -> "W"
            MapDirection.NORTH -> "N"
            MapDirection.SOUTH -> "S"
            else -> ""
        }


    fun isAt(pos: Vector2d?): Boolean {
        return position==pos
    }

    fun move(direction: MoveDirection) {
        when (direction) {
            MoveDirection.RIGHT -> this.direction = this.direction.next()
            MoveDirection.LEFT -> this.direction = this.direction.previous()
            else -> {
                var directionVector: Vector2d = this.direction.toUnitVector()
                if (direction === MoveDirection.BACKWARD) {
                    directionVector = directionVector.opposite()
                }
                val newLocation: Vector2d = position.add(directionVector)
                if (map != null && map!!.canMoveTo(newLocation)) {
                    position = newLocation
                }
                else if (map == null){
                    position = newLocation
                }
            }
        }
    }
}